test_defopt.py

```diff
@@ -68,6 +68,81 @@ def sub_with_dash(*, baz=None):
                        argv=['sub_2', '--baz', '1']), 1)
 
+    def test_multisubcommands(self):
+        def sub(*bar):
+            """:type bar: float"""
+            return bar
+        def subsub_with_dash(*, baz=None):
+            """:type baz: int"""
+            return baz
+        def subsub(*foo):
+            """:type foo: float"""
+            return foo
+        self.assertEqual(
+            defopt.run([sub, {'sub2': [subsub_with_dash, subsub]}],
+                       argv=['sub', '1.1']), (1.1,))
+        self.assertEqual(
+            defopt.run([sub, {'sub2': [subsub_with_dash, subsub]}],
+                       argv=['sub2', 'subsub-with-dash', '--baz', '1']), 1)
+        self.assertEqual(
+            defopt.run([sub, {'sub2': [subsub_with_dash, subsub]}],
+                       argv=['sub2', 'subsub', '1.1']), (1.1,))
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2": [subsub_with_dash, subsub]},
+                       argv=['sub1', '1.2']), (1.2,))
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2": [subsub_with_dash, subsub]},
+                       argv=['sub_2', 'subsub-with-dash', '--baz', '1']), 1)
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2": [subsub_with_dash, subsub]},
+                       argv=['sub_2', 'subsub', '1.5']), (1.5,))
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2":
+                       {'subsub1': subsub_with_dash, 'subsub2': subsub}},
+                       argv=['sub1', '1.2']), (1.2,))
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2":
+                       {'subsub1': subsub_with_dash, 'subsub2': subsub}},
+                       argv=['sub_2', 'subsub1', '--baz', '1']), 1)
+        self.assertEqual(
+            defopt.run({"sub1": sub, "sub_2":
+                       {'subsub1': subsub_with_dash, 'subsub2': subsub}},
+                       argv=['sub_2', 'subsub2', '1.5']), (1.5,))
+
+    def test_multisubcommands_deep(self):
+        def sub(*bar):
+            """:type bar: float"""
+            return bar
+        self.assertEqual(
+            defopt.run({'a': {'b': {'c': {'d': {'e': sub}}}}},
+                       argv=['a', 'b', 'c', 'd', 'e', '1.2']), (1.2,))
+        self.assertEqual(
+            defopt.run({'a': {'b': {'c': {'d': {'e': [sub]}}}}},
+                       argv=['a', 'b', 'c', 'd', 'e', 'sub', '1.2']), (1.2,))
+
+    def test_multisubcommands_mixed(self):
+        def sub(*bar):
+            """:type bar: float"""
+            return bar
+        def subsub_with_dash(*, baz=None):
+            """:type baz: int"""
+            return baz
+        def subsub(*foo):
+            """:type foo: float"""
+            return foo
+        # Note that in this case, the commands are all added at the same level
+        # as it isn't clear what the subcommand name would be for the
+        # dictionary entries.
```

Comment:
In order to move this forward, let's stick to the limited API I propose for now:
- The supported format is a nested dict of names-to-(funcs-or-nested dicts).
- As a special-case, a plain mapping of `{func.__name__: func, ...}` (at any level of nesting) can be replaced by a list of `[func, ...]`.
- In order to allow the "mixed" case, add a `spec` func, basically with the implementation you propose.

The type of `funcs` would then be, I think, the recursive type defined by `T = dict[str, T | Callable] | list[Callable]`.

(I agree that "_1", ... is not particularly helpful.)

Reply:
Ok, I will try to implement this soon.  But I want to make sure I understand exact what subset we want to support.  Is this one allowed?

```
defopt.run([sub1, {'sub2': [subsub1, subsub2]}]
```

I think not, since we're disallowing a mapping within a list altogether, right?

Reply:
This should not be allowed.  Basically, each container (list or dict) introduces a new level, and cannot be at the same level as things above it.